lib/extras/codec_apng.cc

```diff
@@ -658,5 +661,218 @@ Status DecodeImageAPNG(const Span<const uint8_t> bytes,
+class BlobsWriterPNG {
+ public:
+  static Status Encode(const Blobs& blobs, std::vector<png_text>* info,
+                       std::vector<std::string>* strings) {
+    if (!blobs.exif.empty()) {
+      JXL_RETURN_IF_ERROR(EncodeBase16("exif", blobs.exif, info, strings));
+    }
+    if (!blobs.iptc.empty()) {
+      JXL_RETURN_IF_ERROR(EncodeBase16("iptc", blobs.iptc, info, strings));
+    }
+    if (!blobs.xmp.empty()) {
+      JXL_RETURN_IF_ERROR(EncodeBase16("xmp", blobs.xmp, info, strings));
+    }
+    return true;
+  }
+
+ private:
+  static JXL_INLINE char EncodeNibble(const uint8_t nibble) {
+    JXL_ASSERT(nibble < 16);
+    return (nibble < 10) ? '0' + nibble : 'a' + nibble - 10;
+  }
+
+  static Status EncodeBase16(const std::string& type, const PaddedBytes& bytes,
+                             std::vector<png_text>* info,
+                             std::vector<std::string>* strings) {
+    // Encoding: base16 with newline after 72 chars.
+    const size_t base16_size =
+        2 * bytes.size() + DivCeil(bytes.size(), size_t(36)) + 1;
+    std::string base16;
+    base16.reserve(base16_size);
+    for (size_t i = 0; i < bytes.size(); ++i) {
+      if (i % 36 == 0) base16.push_back('\n');
+      base16.push_back(EncodeNibble(bytes[i] >> 4));
+      base16.push_back(EncodeNibble(bytes[i] & 0x0F));
+    }
+    base16.push_back('\n');
+    JXL_ASSERT(base16.length() == base16_size);
+
+    char key[30];
+    snprintf(key, sizeof(key), "Raw profile type %s", type.c_str());
+
+    char header[30];
+    snprintf(header, sizeof(header), "\n%s\n%8" PRIuS, type.c_str(),
+             bytes.size());
+
+    strings->push_back(std::string(key));
+    png_text text;
+    text.key = const_cast<png_charp>(strings->back().data());
+    strings->push_back(std::string(header) + base16);
+    text.text = const_cast<png_charp>(strings->back().data());
+    text.compression = PNG_TEXT_COMPRESSION_zTXt;
+    info->push_back(text);
+    return true;
+  }
+};
+
+Status EncodeImageAPNG(const CodecInOut* io, const ColorEncoding& c_desired,
+                       size_t bits_per_sample, ThreadPool* pool,
+                       PaddedBytes* bytes) {
+  if (bits_per_sample > 8) {
+    bits_per_sample = 16;
+  } else if (bits_per_sample < 8) {
+    // PNG can also do 4, 2, and 1 bits per sample, but it isn't implemented
+    bits_per_sample = 8;
+  }
+
+  size_t count = 0;
+  bool have_anim = io->metadata.m.have_animation;
+  size_t anim_chunks = 0;
+  int W = 0, H = 0;
+
+  for (auto& frame : io->frames) {
+    png_structp png_ptr;
+    png_infop info_ptr;
+
+    png_ptr = png_create_write_struct(PNG_LIBPNG_VER_STRING, NULL, NULL, NULL);
+
+    if (!png_ptr) return JXL_FAILURE("Could not init png encoder");
+
+    info_ptr = png_create_info_struct(png_ptr);
+    if (!info_ptr) return JXL_FAILURE("Could not init png info struct");
+
+#ifndef __GNUC__
+    // gcc produces spurious "variable might be clobbered by ‘longjmp’" errors
```

Comment:
I'm not sure this is a spurious warning - we have types with nontrivial destructors on the same stack frame as where we setjmp, which is *never* a good idea...

Reply:
Right. Maybe we can just skip all the setjmp setting in codec_apng.cc (both for encode and decode), which makes libpng abort on error instead of us catching it and returning failure, but the code is not part of libjxl, only used in tools, so maybe that's OK?

Reply:
Why don't we just put the setjmp in an outer function?

Reply:
I don't see an easy way to do that, but I think it's fine to use the default libpng error handling instead of catching them. It shouldn't happen that you get png _encode_  errors, and if you do, it's not going to make much of a difference if the error gets caught or not. So I'm just removing the setjmp in the encode path.

Reply:
I'm not 100% sure I like this but I am fine with it until proven otherwise :)